sdk/runanywhere-python/tests/test_grammar.py

```
@@ -0,0 +1,586 @@
+    """The set of defined rule names (text left of '::=' on each line)."""
+    names: set[str] = set()
+    for line in grammar.split("\n"):
+        if "::=" not in line:
+            continue
+        names.add(line.split("::=")[0].strip())
+    return names
+
+
+def assert_closed(grammar: str) -> None:
+    """Every referenced identifier must be a defined rule; exactly one root; no dup LHS."""
+    lines = [l for l in grammar.split("\n") if "::=" in l]
```

Ruff (0.15.21)

[error] Ambiguous variable name: `l`

(E741)